### PR TITLE
Bug 1917587: disable Manila operator in case of 404 error

### PR DESCRIPTION
**Pull request by Fedosin**
We have to handle 404 errors from Manila and disable the operator in this case.

### CODE REVIEW

File: pkg/controllers/manila/manila.go [hunk 1]
```diff
@@ -103,6 +103,9 @@ func (c *ManilaController) sync(ctx context.Context, syncCtx factory.SyncContext
 		case *gophercloud.ErrEndpointNotFound:
 			// OpenStack does not support manila, report the operator as disabled
 			return c.setDisabledCondition("This OpenStack cluster does not provide Manila service")
+		case gophercloud.ErrDefault404:
+			// Manila API is modified and not supported by the operator
+			return c.setDisabledCondition("Manila API is modified and not supported by the operator")
```
Comment (mandre) [inline pkg/controllers/manila/manila.go:108]: The thing is that [`GetShareTypes()`](https://github.com/openshift/csi-driver-manila-operator/blob/2d0d211/pkg/controllers/manila/openstack.go#L37) can return many more legitimate errors than `ErrDefault403` or `ErrEndpointNotFound`. As we can see from the logs, we're getting a 404 and so far we don't know what function call is causing it.

I think we need to:
- add more logs to `GetShareTypes()` to identify what function call returns a 404.
- improve the error message when we catch the 404: "Manila API is modified and not supported by the operator" is a supposition and not actual fact.

Think for example of a case where keystone is misconfigured and returns a wrong endpoint for Manila. What should be the error message?

File: pkg/controllers/manila/manila.go [hunk 2]
```diff
@@ -94,14 +95,23 @@ func (c *ManilaController) sync(ctx context.Context, syncCtx factory.SyncContext
 		return nil
 	}
 
+	var err403 gophercloud.ErrDefault403
+	var err404 gophercloud.ErrDefault403
```
Comment (mandre) [inline pkg/controllers/manila/manila.go:99]:
```suggestion
	var err404 gophercloud.ErrDefault404
```

**Review by mandre — CHANGES_REQUESTED** [review]

File: pkg/controllers/manila/manila.go (hunk 2 above)
Comment (Fedosin) [reply pkg/controllers/manila/manila.go]: thanks!

File: pkg/controllers/manila/openstack.go [hunk 3]
```diff
@@ -80,19 +80,19 @@ func (o *openStackClient) GetShareTypes() ([]sharetypes.ShareType, error) {
 
 	err = openstack.Authenticate(provider, *opts)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot authenticate with given credentials: %w", err)
 	}
 
 	client, err := openstack.NewSharedFileSystemV2(provider, gophercloud.EndpointOpts{
```
Comment (jsafrane) [inline pkg/controllers/manila/openstack.go:86]: `GetShareTypes()` does many things. Is it possible that say `404` from `NewSharedFileSystemV2` needs to be handled differently than `404` from e.g. `sharetypes.List` below?

The operator will get disabled with the same message in both cases.

Comment (Fedosin) [reply pkg/controllers/manila/openstack.go:86]: It's a good question. My thoughts are:
- ErrEndpointNotFound can come from NewSharedFileSystemV2 only;
- ErrDefault403 always comes from sharetypes.List, Keystone doesn't return this code for the authentication request;
- ErrDefault404 potentially can be returned both from Keystone and Manila, but frankly speaking it is an absolutely extreme scenario for Keystone and I can't imagine a situation that may lead to this.

File: pkg/controllers/manila/manila.go [hunk 4]
```diff
@@ -94,14 +95,23 @@ func (c *ManilaController) sync(ctx context.Context, syncCtx factory.SyncContext
...
 
+	var err403 gophercloud.ErrDefault403
+	var err404 gophercloud.ErrDefault404
+	var errNoEndpoint *gophercloud.ErrEndpointNotFound
 	shareTypes, err := c.openStackClient.GetShareTypes()
 	if err != nil {
-		switch err.(type) {
-		case gophercloud.ErrDefault403:
+		switch {
+		case errors.As(err, &err403):
```
Comment (bogdando) [inline pkg/controllers/manila/manila.go:104]: you may want to use that IsNotFound wrapper for these as well

Comment (Fedosin) [reply pkg/controllers/manila/manila.go:104]: You know, you right. You never know what to expect from gophercloud.

**Review by mandre — COMMENTED** [review]
/lgtm

**Pull request merged**